### PR TITLE
[fix]: [CI-3384]: [TI] addon check for harnessArgLine tag in pom.xml

### PR DESCRIPTION
**Pull request by hemanthmantri**
[fix]: [CI-3384]: [TI] addon should switch between -DargLine and -DharnessArgLine 

This change introduces a helper method to check the presence of harnessArgLine tag in any XML
file.  If such a tag exists, the -DArgLine will be replaced by -DharnessArgLine

You can use the following comments to re-trigger PR Checks

- Compile: `trigger compile`
- runAeriformCheck: `trigger AeriformCheck`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`

[CI-3384]: https://harness.atlassian.net/browse/CI-3384?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/30464)

### CODE REVIEW

File: product/ci/addon/testintelligence/java/helper.go [hunk 1]
```diff
@@ -32,6 +35,65 @@ func getFiles(path string) ([]string, error) {
...
+
+	// Check each XML file for the harnessArgLine tag.
+	for _, f := range files {
+		absPath, err := filepath.Abs(f)
+		if err == nil {
+			fmt.Println("Checking file: ", absPath)
+			log.Infow(fmt.Sprintf("Checking XML file %s for Harness tag", absPath))
+			found, _ := checkHarnessTag(absPath)
+			if found {
+				foundHarnessTag = true
```
Comment (shivagowda) [inline product/ci/addon/testintelligence/java/helper.go:88]: we can skip checking further once we find the tag
continue

**Comment by shivagowda** [comment]
Thanks for taking care of this @hemanthmantri 
please add unit tests
for go code, add `go` label while raising the PR so that go builds get kicked off.

File: product/ci/addon/testintelligence/java/helper.go [hunk 2]
```diff
@@ -32,6 +35,65 @@ func getFiles(path string) ([]string, error) {
...
+//     bool:  True if the Harness tag exists, false otherwise
+//     error: error if there's any, nil otherwise
+
+func checkHarnessTag(filePath string) (bool, error) {
+	data, err := os.ReadFile(filePath)
+	if err != nil {
+		log.Errorw("Failed to read file at: ", filePath)
+		return false, err
+	}
+	harnessTag := "harnessArgLine" // FIXME: Make this a global constant or a flag.
```
Comment (shivagowda) [inline product/ci/addon/testintelligence/java/helper.go:53]: pls make this as constant

File: product/ci/addon/testintelligence/java/helper.go [hunk 3]
```diff
@@ -32,6 +35,65 @@ func getFiles(path string) ([]string, error) {
...
+//
+// Args:
+//     filePath (string): The absolute path to XML file to be checked.
+//
+// Returns:
+//     bool:  True if the Harness tag exists, false otherwise
+//     error: error if there's any, nil otherwise
+
+func checkHarnessTag(filePath string) (bool, error) {
+	data, err := os.ReadFile(filePath)
```
Comment (vistaarjuneja) [inline product/ci/addon/testintelligence/java/helper.go:48]: We have created a filesystem package which is a mock-able interface to perform file system commands. That helps in unit tests to mock certain behavior. You can search for "filesystem.Filesystem" to see how it gets used

File: product/ci/addon/testintelligence/java/maven.go [hunk 4]
```diff
@@ -48,14 +49,20 @@ func (m *mavenRunner) GetCmd(ctx context.Context, tests []types.RunnableTest, us
...
+
 	if s != nil {
 		// If user args are present, move them to instrumentation
 		userArgs = re.ReplaceAllString(userArgs, "")                        // Remove from arg
 		instrArg = fmt.Sprintf("\"%s %s\"", strings.Join(s, " "), agentArg) // Add to instrumentation
 	}
 	if runAll {
 		// Run all the tests
-		return strings.TrimSpace(fmt.Sprintf("%s -am -DargLine=%s %s", mavenCmd, instrArg, userArgs)), nil
+		return strings.TrimSpace(fmt.Sprintf("%s -am -D%s=%s %s", mavenCmd, argLineStr, instrArg, userArgs)), nil
```
Comment (vistaarjuneja) [inline product/ci/addon/testintelligence/java/maven.go:65]: Unit tests would have to be updated as well

File: product/ci/addon/testintelligence/java/helper.go [hunk 5]
```diff
@@ -32,6 +35,65 @@ func getFiles(path string) ([]string, error) {
...
+			log.Infow(fmt.Sprintf("Checking XML file %s for Harness tag", absPath))
+			found, _ := checkHarnessTag(absPath)
+			if found {
+				foundHarnessTag = true
+			}
+		} else {
+			log.Errorw("Failed to absolute path. Error", err)
+		}
+	}
+	return foundHarnessTag, err
```
Comment (vistaarjuneja) [inline product/ci/addon/testintelligence/java/helper.go:94]: In golang, generally the convention is to check whether error is nil, and if not return nil explicitly. eg:
`if err != nil {
    return ret, err
}
return ret, nil`

File: product/ci/addon/testintelligence/java/helper.go [hunk 6]
```diff
@@ -32,6 +35,65 @@ func getFiles(path string) ([]string, error) {
...
+}
+
+// Scan all XML files in the workspace and detect if any of them have harnessArgLine tag.
+
+func DetectHarnessTag() (bool, error) {
+	// Get all the XML files from $HARNESS_WORKSPACE
+	wp, _ := GetWrkspcPath()
+	files, _ := getFiles(fmt.Sprintf("%s/**/*.xml", wp))
+	foundHarnessTag := false
+	err := nil
```
Comment (vistaarjuneja) [inline product/ci/addon/testintelligence/java/helper.go:78]: Prefer to just declare `var err error` here and use the check in the next comment

File: product/ci/addon/testintelligence/java/helper.go [hunk 7]
```diff
@@ -32,6 +35,65 @@ func getFiles(path string) ([]string, error) {
...
+
+	// Check each XML file for the harnessArgLine tag.
+	for _, f := range files {
+		absPath, err := filepath.Abs(f)
```
Comment (vistaarjuneja) [inline product/ci/addon/testintelligence/java/helper.go:82]: Alternative:
```
if err != nil {
    // log error
   continue
}
...
```

This avoids two code blocks which is generally harder to follow

File: product/ci/addon/testintelligence/java/helper.go [hunk 8]
```diff
@@ -32,6 +35,65 @@ func getFiles(path string) ([]string, error) {
...
+		if err == nil {
+			fmt.Println("Checking file: ", absPath)
+			log.Infow(fmt.Sprintf("Checking XML file %s for Harness tag", absPath))
+			found, _ := checkHarnessTag(absPath)
+			if found {
+				foundHarnessTag = true
+			}
+		} else {
+			log.Errorw("Failed to absolute path. Error", err)
```
Comment (vistaarjuneja) [inline product/ci/addon/testintelligence/java/helper.go:91]: Prefer keeping log error strings starting with lowercase

File: product/ci/addon/testintelligence/java/helper.go [hunk 9]
```diff
@@ -32,6 +35,65 @@ func getFiles(path string) ([]string, error) {
...
+	wp, _ := GetWrkspcPath()
+	files, _ := getFiles(fmt.Sprintf("%s/**/*.xml", wp))
+	foundHarnessTag := false
+	err := nil
+
+	// Check each XML file for the harnessArgLine tag.
+	for _, f := range files {
+		absPath, err := filepath.Abs(f)
+		if err == nil {
+			fmt.Println("Checking file: ", absPath)
```
Comment (vistaarjuneja) [inline product/ci/addon/testintelligence/java/helper.go:84]: Remove this line

File: product/ci/addon/testintelligence/java/helper.go [hunk 10]
```diff
@@ -32,6 +35,65 @@ func getFiles(path string) ([]string, error) {
...
+//     filePath (string): The absolute path to XML file to be checked.
+//
+// Returns:
+//     bool:  True if the Harness tag exists, false otherwise
+//     error: error if there's any, nil otherwise
+
+func checkHarnessTag(filePath string) (bool, error) {
+	data, err := os.ReadFile(filePath)
+	if err != nil {
+		log.Errorw("Failed to read file at: ", filePath)
```
Comment (vistaarjuneja) [inline product/ci/addon/testintelligence/java/helper.go:50]: Prefer to start error strings with lower case

Comment (vistaarjuneja) [reply product/ci/addon/testintelligence/java/helper.go:50]: log the Error as well: `log.Errorw("failed to read file", "file_path", filePath, zap.Error(err))`

**Comment by vistaarjuneja** [comment]
@hemanthmantri  - these changes would need to be cherry picked in the new lite engine as well: https://github.com/harness/lite-engine

**Comment by vistaarjuneja** [comment]
For golang PRs, adding the `go` label will trigger the Go build & skip some of the java tests which are run

File: product/ci/addon/testintelligence/java/helper.go (hunk 10 above)
Comment (hemanthmantri) [reply product/ci/addon/testintelligence/java/helper.go]: Done

File: product/ci/addon/testintelligence/java/helper.go (hunk 2 above)
Comment (hemanthmantri) [reply product/ci/addon/testintelligence/java/helper.go]: Done

File: product/ci/addon/testintelligence/java/helper.go (hunk 6 above)
Comment (hemanthmantri) [reply product/ci/addon/testintelligence/java/helper.go]: Done

File: product/ci/addon/testintelligence/java/helper.go (hunk 7 above)
Comment (hemanthmantri) [reply product/ci/addon/testintelligence/java/helper.go:97]: Done

File: product/ci/addon/testintelligence/java/helper.go (hunk 1 above)
Comment (hemanthmantri) [reply product/ci/addon/testintelligence/java/helper.go]: Done. I think you meant break.

File: product/ci/addon/testintelligence/java/helper.go (hunk 8 above)
Comment (hemanthmantri) [reply product/ci/addon/testintelligence/java/helper.go]: Done.

File: product/ci/addon/testintelligence/java/helper.go (hunk 5 above)
Comment (hemanthmantri) [reply product/ci/addon/testintelligence/java/helper.go]: Done.

File: product/ci/addon/testintelligence/java/maven.go (hunk 4 above)
Comment (hemanthmantri) [reply product/ci/addon/testintelligence/java/maven.go:66]: Done.

File: product/ci/addon/testintelligence/java/helper.go (hunk 3 above)
Comment (hemanthmantri) [reply product/ci/addon/testintelligence/java/helper.go]: Done.

File: product/ci/addon/testintelligence/java/helper.go (hunk 9 above)
Comment (hemanthmantri) [reply product/ci/addon/testintelligence/java/helper.go]: Done.

**Review by hemanthmantri — COMMENTED** [review]
Addressed all comments and added unit tests. Please review again.

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30464#-:-Mv5Pq-05z1XDHHfa8gu:b-lzk5et)* status: 8 file(s) to review. 10. unresolved discussions. 0 of 1 approvals obtained.

File: product/ci/addon/testintelligence/java/helper.go [hunk 11]
```diff
@@ -32,6 +36,83 @@ func getFiles(path string) ([]string, error) {
...
+//   fs  (FileSystem): Filesystem interface for File IO.
+//
+// Returns:
+//  bool : True if harnessArgLine is present in any of the XML files. False otherwise.
+//  error: Error if there's one, nil otherwise.
+
+func DetectHarnessTag(log *zap.SugaredLogger, fs filesystem.FileSystem) (bool, error) {
+	// Get all the XML files from $HARNESS_WORKSPACE
+	wp, _ := getWorkspace()
+	files, _ := getFiles(fmt.Sprintf("%s/**/*.xml", wp))
```
Comment (shivagowda) [inline product/ci/addon/testintelligence/java/helper.go:91]: it would be safe to scan only pom.xml
shall we change this?  missed this comment earlier

**Review by vistaarjuneja — APPROVED** [review]
:lgtm:

> Reviewed 8 of 8 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30464#-:-MvHkpchE62Kv1MPNqlD:bnfp4nl)* status: 10. unresolved discussions.

**Review by vistaarjuneja — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30464)* status: 2. unresolved discussions.

File: product/ci/addon/testintelligence/java/helper.go (hunk 11 above)
Comment (hemanthmantri) [reply product/ci/addon/testintelligence/java/helper.go]: Done in updated patchset.

**Comment by hemanthmantri** [comment]
Closing this PR. The fix is being done via https://github.com/harness/harness-core/pull/30721

**Pull request closed without merging**